### ISSUE TITLE
OperationTimeoutException AddCacheConfigOperation invocation failed to complete due to operation-heartbeat-timeout

### ISSUE THREAD

**Issue by Danny-Hazelcast** [hazelcast/hazelcast#13216]
```
fail HzMember3HZ _create-use-destroy_createUseDistDestroy_cacheBak1HD-ncHD hzcmd.cache.CreateUseDestroy threadId=3 com.hazelcast.core.OperationTimeoutException: AddCacheConfigOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2018-05-29 20:46:49.514. Start time: 2018-05-29 20:26:49.325. Total elapsed time: 1200189 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2018-05-29 20:46:21.571. Invocation{op=com.hazelcast.cache.impl.operation.AddCacheConfigOperation{serviceName='hz:impl:cacheService', identityHash=218089881, partitionId=-1, replicaIndex=0, callId=-99, invocationTime=1527625609319 (2018-05-29 20:26:49.319), waitTimeout=-1, callTimeout=600000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=600000, firstInvocationTimeMs=1527625609325, firstInvocationTime='2018-05-29 20:26:49.325', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.0.0.211]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=Connection[id=2, /10.0.0.129:36843->/10.0.0.211:5701, endpoint=[10.0.0.211]:5701, alive=true, type=MEMBER]}
```

https://hazelcast-l337.ci.cloudbees.com/view/stable/job/stable-All/91/console

http://54.82.84.143/~jenkins/workspace/stable-All/3.10.2/2018_05_29-17_00_14/destroy/create-use-destroy-HDnearcache

/disk1/jenkins/workspace/stable-All/3.10.2/2018_05_29-17_00_14/destroy/create-use-destroy-HDnearcache

```
./output/HZ/HzClient3HZ/exception.txt
./output/HZ/HzMember2HZ/exception.txt
./output/HZ/HzMember4HZ/exception.txt
./output/HZ/HzClient4HZ/exception.txt
./output/HZ/HzMember1HZ/exception.txt
./output/HZ/HzMember3HZ/exception.txt
./output/HZ/HzClient2HZ/exception.txt
./output/HZ/HzClient1HZ/exception.txt
```

```
cat ./output/HZ/HzClient3HZ/exception.txt
com.hazelcast.core.OperationTimeoutException: RegistrationOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2018-05-29 20:46:49.538. Start time: 2018-05-29 20:26:49.369. Total elapsed time: 1200171 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2018-05-29 20:46:21.567. Invocation{op=com.hazelcast.spi.impl.eventservice.impl.operations.RegistrationOperation{serviceName='null', identityHash=1025791524, partitionId=-1, replicaIndex=0, callId=-129, invocationTime=1527625609369 (2018-05-29 20:26:49.369), waitTimeout=-1, callTimeout=600000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=600000, firstInvocationTimeMs=1527625609369, firstInvocationTime='2018-05-29 20:26:49.369', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.0.0.211]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=Connection[id=5, /10.0.0.110:5701->/10.0.0.211:53256, endpoint=[10.0.0.211]:5701, alive=true, type=MEMBER]}
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.newOperationTimeoutException(InvocationFuture.java:164)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:106)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:250)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:127)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:250)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:96)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:33)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:162)
	at com.hazelcast.client.spi.ProxyManager.initialize(ProxyManager.java:436)
	at com.hazelcast.client.spi.ProxyManager.initializeWithRetry(ProxyManager.java:395)
	at com.hazelcast.client.spi.ProxyManager.getOrCreateProxy(ProxyManager.java:316)
	at com.hazelcast.client.impl.HazelcastClientInstanceImpl.getDistributedObject(HazelcastClientInstanceImpl.java:707)
	at com.hazelcast.client.impl.HazelcastClientInstanceImpl.getMap(HazelcastClientInstanceImpl.java:511)
	at com.hazelcast.client.impl.HazelcastClientProxy.getMap(HazelcastClientProxy.java:120)
	at hzcmd.map.CreateUseDestroy.timeStep(CreateUseDestroy.java:11)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)
	at remote.bench.BenchThread.call(BenchThread.java:38)
	at remote.bench.BenchThread.call(BenchThread.java:12)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:473)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
	at java.lang.Thread.run(Thread.java:748)
```

some log snippets.
```
./output/HZ/HzMember3HZ/out.txt:WARNING: [10.0.0.129]:5701 [HZ] [3.10.2] Can't fetch invalidation meta-data [TimeoutException] MapGetInvalidationMetaDataOperation failed to complete within 1 MINUTES. Invocation{op=com.hazelcast.map.impl.operation.MapGetInvalidationMetaDataOperation{serviceName='hz:impl:mapService', identityHash=2099593127, partitionId=-1, replicaIndex=0, callId=124, invocationTime=1527625668933 (2018-05-29 20:27:48.933), waitTimeout=-1, callTimeout=600000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=600000, firstInvocationTimeMs=1527625668939, firstInvocationTime='2018-05-29 20:27:48.939', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.0.0.110]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=Connection[id=4, /10.0.0.129:5701->/10.0.0.110:56444, endpoint=[10.0.0.110]:5701, alive=true, type=MEMBER]}
```

```
./output/HZ/HzMember1HZ/out.txt:WARNING: [10.0.0.171]:5701 [HZ] [3.10.2] Slow operation detected: com.hazelcast.spi.impl.proxyservice.impl.operations.DistributedObjectDestroyOperati
on
./output/HZ/HzMember1HZ/out.txt:WARNING: [10.0.0.171]:5701 [HZ] [3.10.2] Slow operation detected: com.hazelcast.spi.impl.proxyservice.impl.operations.DistributedObjectDestroyOperati
on (2 invocations)
```

**Comment by Danny-Hazelcast** [comment]
test
https://github.com/hazelcast/hzCmd-bench/tree/zeta/lab/hz/stable/destroy/create-use-destroy-HDnearcache


https://github.com/hazelcast/hzCmd-bench/blob/zeta/lab/hz/stable/destroy/create-use-destroy-HDnearcache/createUseDestroy

just create use and destroy HD map and cache

xml config
https://github.com/hazelcast/hzCmd-bench/tree/zeta/lab/hz/stable/destroy/create-use-destroy-HDnearcache/config-hz

**Comment by mmedenjak** [comment]
Possibly a duplicate of: https://github.com/hazelcast/hazelcast/issues/13873

**Comment by mmedenjak** [comment]
@Danny-Hazelcast can you retest with latest master?

**Comment by Danny-Hazelcast** [comment]
(base) [jenkins@ip-10-72-134-107 create-use-destroy-HDnearcache]$ pwd
/disk1/jenkins/workspace/stable-All/4.0-SNAPSHOT/2019_08_19-14_32_46/destroy/create-use-destroy-HDnearcache

http://54.147.27.51/~jenkins/workspace/stable-All/4.0-SNAPSHOT/2019_08_19-14_32_46/destroy/create-use-destroy-HDnearcache/out.txt

```
HzMember1HZ _create-use-destroy_mapBak1HD-ncHD hzcmd.map.CreateUseDestroy threadId=0 com.hazelcast.core.OperationTimeoutException: RegistrationOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2019-08-19 22:33:27.648. Start time: 2019-08-19 22:13:27.064. Total elapsed time: 1200584 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2019-08-19 22:33:03.720. Invocation{op=com.hazelcast.spi.impl.eventservice.impl.operations.RegistrationOperation{serviceName='null', identityHash=552300066, partitionId=-1, replicaIndex=0, callId=-109, invocationTime=1566252807059 (2019-08-19 22:13:27.059), waitTimeout=-1, callTimeout=600000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=600000, firstInvocationTimeMs=1566252807064, firstInvocationTime='2019-08-19 22:13:27.064', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.0.0.28]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=Connection[id=3, /10.0.0.35:54431->/10.0.0.28:5701, qualifier=null, endpoint=[10.0.0.28]:5701, alive=true, type=MEMBER]}
```

**Comment by Danny-Hazelcast** [comment]
```(base) [jenkins@ip-10-72-134-107 create-use-destroy-HDnearcache]$ more output/HZ/HzMember1HZ/exception.txt
com.hazelcast.core.OperationTimeoutException: RegistrationOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2019-08-19 22:33:27.648. Start ti
me: 2019-08-19 22:13:27.064. Total elapsed time: 1200584 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2019-08-19 22:33:03.720. Invocation{op=com.hazelc
ast.spi.impl.eventservice.impl.operations.RegistrationOperation{serviceName='null', identityHash=552300066, partitionId=-1, replicaIndex=0, callId=-109, invocationTime=1566252807059
 (2019-08-19 22:13:27.059), waitTimeout=-1, callTimeout=600000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=600000, firstInvocationTimeMs=1566252807064, firs
tInvocationTime='2019-08-19 22:13:27.064', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.0.0.28]:5701, pendingResponse={VOID}, backupsAcksExpected=0
, backupsAcksReceived=0, connection=Connection[id=3, /10.0.0.35:54431->/10.0.0.28:5701, qualifier=null, endpoint=[10.0.0.28]:5701, alive=true, type=MEMBER]}
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.newOperationTimeoutException(InvocationFuture.java:163)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:106)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:251)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
```

**Comment by vbekiaris** [comment]
@Danny-Hazelcast can we get a re-run with `-Dhazelcast.slow.operation.detector.stacktrace.logging.enabled`, diagnostics enabled and (if possible) periodic thread dumps every 30 seconds?

**Comment by mmedenjak** [comment]
@Danny-Hazelcast ☝️

**Comment by Danny-Hazelcast** [comment]
a run http://54.147.27.51/~jenkins/workspace/stable-x3/4.0-SNAPSHOT/2019_12_12-04_26_24/create-use-destroy-HDnearcache/

gives member OOME,  so over a time the issue has changed a bit ? 

```

00:09:25.213 /disk1/jenkins/workspace/stable-x3/4.0-SNAPSHOT/2019_12_12-04_26_24/create-use-destroy-HDnearcache Failed
00:09:25.218 fail HzClient1HZ _create-use-destroy_cacheBak1HD-ncHD hzcmd.cache.CreateUseDestroy threadId=1 com.hazelcast.spi.exception.TargetDisconnectedException: Heartbeat timed out [0m
00:09:25.378 http://54.147.27.51/~jenkins/workspace/stable-x3/4.0-SNAPSHOT/2019_12_12-04_26_24/create-use-destroy-HDnearcache
00:09:25.378  
00:10:34.284 ./create-use-destroy-HDnearcache/output/HZ/HzMember2HZ/HzMember2HZ.hprof
00:10:34.316 ./create-use-destroy-HDnearcache/output/HZ/HzClient1HZ/exception.txt


```

**Comment by Danny-Hazelcast** [comment]
<img width="893" alt="Screen Shot 2019-12-12 at 13 29 09" src="https://user-images.githubusercontent.com/5988678/70685179-89023a80-1ce3-11ea-9341-ce6128759260.png">
<img width="804" alt="Screen Shot 2019-12-12 at 13 29 18" src="https://user-images.githubusercontent.com/5988678/70685181-89023a80-1ce3-11ea-9fcd-fdb23b5fd5d5.png">

**Comment by vbekiaris** [comment]
A single event executor's queue has 165K entries; 110K of them are `RemoveEventProcessor` instances. I sampled the `Packet`'s payload in a few of them and I found `BatchNearCacheInvalidation`s, each being ~12K in size.

In member2's heap dump there are overall 110015 instances of `RemoteEventProcessor`; 110014 of those instances are in the specific executor's queue. This seems in line with the test configuration which has a max of 1 -> all create-use-destroy actions occur on a single map and a single cache.

The OOME is either due to the single member not being able to keep up with the rate of invalidation events or due to event queue processing being blocked (eg one possibility could be proxy destruction event resulting in fan-out of remote partition/generic operations). In order to find out, @Danny-Hazelcast can we re-run the test with diagnostics and periodic thread dumps enabled?

**Comment by Danny-Hazelcast** [comment]
can we constrain the size of an event executor's queue ?

http://jenkins.hazelcast.com/view/stable/job/stable-x1/51/console


00:21:58.918 /disk1/jenkins/workspace/stable-x1/4.0-SNAPSHOT/2019_12_17-09_31_12/create-use-destroy-HDnearcache Failed
00:21:58.924 fail HzClient1HZ _create-use-destroy_cacheBak1HD-ncHD hzcmd.cache.CreateUseDestroy threadId=2 com.hazelcast.core.OperationTimeoutException: AddCacheConfigOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2019-12-17 09:53:06.027. Start time: 2019-12-17 09:33:05.541. Total elapsed time: 1200486 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2019-12-17 09:52:40.128. Invocation{op=com.hazelcast.cache.impl.operation.AddCacheConfigOperation{serviceName='hz:impl:cacheService', identityHash=145001814, partitionId=-1, replicaIndex=0, callId=210, invocationTime=1576575185538 (2019-12-17 09:33:05.538), waitTimeout=-1, callTimeout=600000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=600000, firstInvocationTimeMs=1576575185541, firstInvocationTime='2019-12-17 09:33:05.541', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.0.0.23]:5701, pendingResponse={VOID}, backupsAcksExpected=-1, backupsAcksReceived=0, connection=Connection[id=1, /10.0.0.112:45065->/10.0.0.23:5701, qualifier=null, endpoint=[10.0.0.23]:5701, alive=true, connectionType=MEMBER]} [0m
00:21:59.034 http://54.147.27.51/~jenkins/workspace/stable-x1/4.0-SNAPSHOT/2019_12_17-09_31_12/create-use-destroy-HDnearcache

**Comment by vbekiaris** [comment]
yes, use property [`hazelcast.event.queue.capacity`](https://github.com/hazelcast/hazelcast/blob/573f6d1b482f59c1633eccb39020746b04b3df90/hazelcast/src/main/java/com/hazelcast/spi/properties/ClusterProperty.java#L171-L172). Its default size is 1_000_000.

**Comment by vbekiaris** [comment]
Notice that event queue capacity is split among event worker threads. Default number of event worker threads is 5, so maximum default capacity per worker is 200_000. In the hprof [reported here](https://github.com/hazelcast/hazelcast/issues/13216#issuecomment-564853158) a single event worker thread's queue had 165K events so we hit OOME before reaching capacity.

**Comment by Danny-Hazelcast** [comment]
using

  <property name="hazelcast.event.queue.capacity">100000</property>

http://jenkins.hazelcast.com/view/stable/job/stable-x1/54/console

```
fail HzMember2HZ _create-use-destroy_mapBak1HD-ncHD hzcmd.map.CreateUseDestroy threadId=3 com.hazelcast.core.OperationTimeoutException: RegistrationOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2019-12-18 08:55:08.376. Start time: 2019-12-18 08:35:08.114. Total elapsed time: 1200263 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2019-12-18 08:54:38.580. Invocation{op=com.hazelcast.spi.impl.eventservice.impl.operations.RegistrationOperation{serviceName='null', identityHash=440457735, partitionId=-1, replicaIndex=0, callId=260, invocationTime=1576658108113 (2019-12-18 08:35:08.113), waitTimeout=-1, callTimeout=600000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=600000, firstInvocationTimeMs=1576658108114, firstInvocationTime='2019-12-18 08:35:08.114', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.0.0.114]:5701, pendingResponse={VOID}, backupsAcksExpected=-1, backupsAcksReceived=0, connection=Connection[id=1, /10.0.0.225:5701->/10.0.0.114:48614, qualifier=null, endpoint=[10.0.0.114]:5701, alive=true, connectionType=MEMBER]}
```

 http://54.147.27.51/~jenkins/workspace/stable-x1/4.0-SNAPSHOT/2019_12_18-08_33_09/create-use-destroy-HDnearcache

**Comment by vbekiaris** [comment]
This latest run looks like a deadlock. Each member reports 2 slow invocations with this stack trace:
```
java.lang.Object.wait(Native Method)
java.lang.Object.wait(Object.java:502)
com.hazelcast.spi.impl.proxyservice.impl.DistributedObjectFuture.waitUntilSetAndInitialized(DistributedObjectFuture.java:65)
com.hazelcast.spi.impl.proxyservice.impl.DistributedObjectFuture.get(DistributedObjectFuture.java:44)
com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.destroyProxy(ProxyRegistry.java:262)
com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.destroyLocalDistributedObject(ProxyServiceImpl.java:185)
com.hazelcast.spi.impl.proxyservice.impl.operations.DistributedObjectDestroyOperation.run(DistributedObjectDestroyOperation.java:45)
com.hazelcast.spi.impl.operationservice.Operation.call(Operation.java:184)
com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:221)
com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:210)
com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:434)
com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:157)
com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:127)com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:114)]
```

Seems proxy destruction is stuck waiting on same proxy's initialization.

**Comment by Danny-Hazelcast** [comment]
http://54.147.27.51/~jenkins/workspace/stable-x1/4.0-SNAPSHOT/2019_12_18-11_28_43/create-use-destroy-HDnearcache/

has j stack traces taken during the run

**Comment by vbekiaris** [comment]
Thanks @Danny-Hazelcast . Summary of the investigation is that there is failure to make any progress. The hzBench tasks involved in this test include:
 * Cache proxy creation which fans out cache config creation operations on all members
 * Map proxy with near cache creation, which involves registering the invalidation listener -> sends operation to register listener on all members
 * Proxy destruction which [blocks waiting on the proxy to be created and initialized](https://github.com/hazelcast/hazelcast/blob/218e195e8c0ccce9f4ea4686b1ea8775e5f3db55/hazelcast/src/main/java/com/hazelcast/spi/impl/proxyservice/impl/ProxyRegistry.java#L262) before proceeding with destruction.

Each member in the test has 2 generic operation threads. Both generic operation threads in members 1 & 3 and 1 generic op thread in member 2 are blocked in a `DistributedObjectDestroyOperation` waiting forever on getting a not-yet-initialized proxy, resulting in operation time out exceptions.

Probably it makes sense that if a user requests destruction of proxy whose `DistributedObjectFuture` is not yet completed, then, instead of blocking forever, the `DistributedObjectFuture` should be completed exceptionally with a `DistributedObjectDestroyedException`.

**Comment by Danny-Hazelcast** [comment]
so client 5 failed with a NOOME, while attempting to clear the near cache ?
SEVERE: hz.client_0 [HZ] [4.0-SNAPSHOT] hz.client_0.event-2 caught an exception while processing:com.hazelcast.client.impl.spi.impl.listener.ClientListenerServiceImpl$ClientEventProcessor@59686061
com.hazelcast.memory.NativeOutOfMemoryError: System allocations limit exceeded! Limit: 2048 KB, usage: 2048 KB, requested: 16 bytes
	at com.hazelcast.internal.memory.AbstractPoolingMemoryManager$MetadataMemoryAllocator.checkSize(AbstractPoolingMemoryManager.java:467)
	at com.hazelcast.internal.memory.AbstractPoolingMemoryManager$MetadataMemoryAllocator.allocate(AbstractPoolingMemoryManager.java:452)
	at com.hazelcast.internal.elastic.queue.LongLinkedBlockingQueue.newNode(LongLinkedBlockingQueue.java:56)
	at com.hazelcast.internal.elastic.queue.LongLinkedBlockingQueue.offer(LongLinkedBlockingQueue.java:95)
	at com.hazelcast.internal.memory.GlobalPoolingMemoryManager$GlobalAddressQueue.release(GlobalPoolingMemoryManager.java:581)
	at com.hazelcast.internal.memory.AbstractPoolingMemoryManager.releaseInternal(AbstractPoolingMemoryManager.java:377)
	at com.hazelcast.internal.memory.AbstractPoolingMemoryManager.free(AbstractPoolingMemoryManager.java:109)
	at com.hazelcast.internal.serialization.impl.EnterpriseSerializationServiceV1.disposeDataInternal(EnterpriseSerializationServiceV1.java:191)
	at com.hazelcast.internal.serialization.impl.EnterpriseSerializationServiceV1.disposeData(EnterpriseSerializationServiceV1.java:181)
	at com.hazelcast.internal.hidensity.impl.DefaultHiDensityRecordProcessor.disposeData(DefaultHiDensityRecordProcessor.java:145)
	at com.hazelcast.internal.elastic.map.BinaryElasticHashMap$SlotIter.removeInternal(BinaryElasticHashMap.java:850)
	at com.hazelcast.internal.elastic.map.BinaryElasticHashMap$KeyIter.removeInternal(BinaryElasticHashMap.java:943)
	at com.hazelcast.internal.elastic.map.BinaryElasticHashMap$SlotIter.remove(BinaryElasticHashMap.java:833)
	at com.hazelcast.internal.elastic.map.BinaryElasticHashMap.clear(BinaryElasticHashMap.java:1177)
	at com.hazelcast.internal.hidensity.impl.DefaultHiDensityRecordMap.clear(DefaultHiDensityRecordMap.java:103)
	at com.hazelcast.internal.nearcache.impl.store.AbstractNearCacheRecordStore.clear(AbstractNearCacheRecordStore.java:370


It seems native memory metadata are allocated but never reclaimed

**Comment by mmedenjak** [comment]
Might be because we allocate some space even if we don't use the structure. See https://github.com/hazelcast/hazelcast-enterprise/issues/1477#issuecomment-311053531

**Comment by vbekiaris** [comment]
The NOOME was encountered on a run on branch https://github.com/hazelcast/hazelcast/pull/16297 that resolved the infinite blocking during proxy destruction issue. Now I need to think a bit more about other consequences this change in proxy destruction might have and add tests to the PR.

**Comment by qq296629801** [comment]
java.util.concurrent.ExecutionException: com.hazelcast.core.OperationTimeoutException: GetConnectedClientsOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2025-02-24 09:06:47.357. Start time: 2025-02-24 09:04:46.818. Total elapsed time: 120540 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2025-02-24 09:05:57.851. Invocation{op=com.hazelcast.client.impl.operations.GetConnectedClientsOperation{serviceName='hz:core:clientEngine', identityHash=685925679, partitionId=-1, replicaIndex=0, callId=-5, invocationTime=1740359086790 (2025-02-24 09:04:46.790), waitTimeout=-1, callTimeout=60000, tenantControl=com.hazelcast.spi.impl.tenantcontrol.NoopTenantControl@0}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1740359086818, firstInvocationTime='2025-02-24 09:04:46.818', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 08:00:00.000', targetAddress=[100.81.112.79]:5701, targetMember=Member [100.81.112.79]:5701 - 40064f8d-8e1c-4b37-bd01-ee7280bfcd86, memberListVersion=2202, pendingResponse={VOID}, backupsAcksExpected=-1, backupsAcksReceived=0, connection=Connection[id=6, /100.70.183.5:42659->/100.81.112.79:5701, qualifier=null, endpoint=[100.81.112.79]:5701, remoteUuid=40064f8d-8e1c-4b37-bd01-ee7280bfcd86, alive=false, connectionType=MEMBER, planeIndex=0]}
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.returnOrThrowWithGetConventions(InvocationFuture.java:121)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrowIfException(InvocationFuture.java:100)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:618)
	at com.hazelcast.client.impl.ClientEngineImpl.getClientsInCluster(ClientEngineImpl.java:496)
	at com.hazelcast.client.impl.ClientEngineImpl.getActiveClientsInCluster(ClientEngineImpl.java:465)
	at com.hazelcast.internal.util.phonehome.ClientInfoCollector.forEachMetric(ClientInfoCollector.java:120)
	at com.hazelcast.internal.util.phonehome.PhoneHome.createParameters(PhoneHome.java:160)
	at com.hazelcast.internal.util.phonehome.PhoneHome.phoneHome(PhoneHome.java:149)
	at com.hazelcast.internal.util.phonehome.PhoneHome.lambda$check$0(PhoneHome.java:97)
	at com.hazelcast.spi.impl.executionservice.impl.DelegateAndSkipOnConcurrentExecutionDecorator$DelegateDecorator.run(DelegateAndSkipOnConcurrentExecutionDecorator.java:77)
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:111)
Caused by: com.hazelcast.core.OperationTimeoutException: GetConnectedClientsOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2025-02-24 09:06:47.357. Start time: 2025-02-24 09:04:46.818. Total elapsed time: 120540 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2025-02-24 09:05:57.851. Invocation{op=com.hazelcast.client.impl.operations.GetConnectedClientsOperation{serviceName='hz:core:clientEngine', identityHash=685925679, partitionId=-1, replicaIndex=0, callId=-5, invocationTime=1740359086790 (2025-02-24 09:04:46.790), waitTimeout=-1, callTimeout=60000, tenantControl=com.hazelcast.spi.impl.tenantcontrol.NoopTenantControl@0}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1740359086818, firstInvocationTime='2025-02-24 09:04:46.818', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 08:00:00.000', targetAddress=[100.81.112.79]:5701, targetMember=Member [100.81.112.79]:5701 - 40064f8d-8e1c-4b37-bd01-ee7280bfcd86, memberListVersion=2202, pendingResponse={VOID}, backupsAcksExpected=-1, backupsAcksReceived=0, connection=Connection[id=6, /100.70.183.5:42659->/100.81.112.79:5701, qualifier=null, endpoint=[100.81.112.79]:5701, remoteUuid=40064f8d-8e1c-4b37-bd01-ee7280bfcd86, alive=false, connectionType=MEMBER, planeIndex=0]}
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.newOperationTimeoutException(InvocationFuture.java:203)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.toExceptionalResult(InvocationFuture.java:167)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:131)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrowIfException(InvocationFuture.java:99)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:618)
	at com.hazelcast.client.impl.ClientEngineImpl.getClientsInCluster(ClientEngineImpl.java:496)
	at com.hazelcast.client.impl.ClientEngineImpl.getActiveClientsInCluster(ClientEngineImpl.java:465)
	at com.hazelcast.internal.util.phonehome.ClientInfoCollector.forEachMetric(ClientInfoCollector.java:120)
	at com.hazelcast.internal.util.phonehome.PhoneHome.createParameters(PhoneHome.java:160)
	at com.hazelcast.internal.util.phonehome.PhoneHome.phoneHome(PhoneHome.java:149)
	at com.hazelcast.internal.util.phonehome.PhoneHome.lambda$check$0(PhoneHome.java:97)
	at com.hazelcast.spi.impl.executionservice.impl.DelegateAndSkipOnConcurrentExecutionDecorator$DelegateDecorator.run(DelegateAndSkipOnConcurrentExecutionDecorator.java:77)
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:111)
	at ------ submitted from ------.()
	at com.hazelcast.internal.util.ExceptionUtil.cloneExceptionWithFixedAsyncStackTrace(ExceptionUtil.java:348)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.returnOrThrowWithGetConventions(InvocationFuture.java:112)
	... 15 more
